File: dungeons-26.1.2/src/main/java/fishmod/features/CompactTab.java
package fishmod.features;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.components.PlayerFaceExtractor;
import net.minecraft.client.multiplayer.PlayerInfo;
import net.minecraft.network.chat.Component;

/**
 * Compact custom tab list. Hypixel packs every column (Players / Info / Trophy Frogs / Active
 * Effects) into the player-list entries, ordered column-major by their {@code !A-a}/{@code !B-a}
 * sort keys — so we render those entries verbatim into a translucent panel rather than re-deriving
 * data. Adds a header stat bar (Players / Server / TPS / Ping) and a footer. Opt-in via config.
 */
public final class CompactTab {
    private CompactTab() {}

    private static final int BG_RGB  = 0x0B0D13;

    /** Panel background alpha derived from the FishSettings slider (0..100). */
    private static int bgPanel() {
        int pct = Math.max(0, Math.min(100, fishmod.utils.config.values.FishSettings.compactTabOpacity));
        int a = (int) Math.round(pct * 2.55); // 0..255
        return (a << 24) | BG_RGB;
    }
    /** Header strip uses ~45% of the panel alpha so it reads as a lighter band. */
    private static int bgHead() {
        int pct = Math.max(0, Math.min(100, fishmod.utils.config.values.FishSettings.compactTabOpacity));
        int a = (int) Math.round(pct * 2.55 * 0.45); // 0..115
        return (a << 24) | BG_RGB;
    }
    private static final int BORDER  = 0x66303440;
    private static final int DIVIDER = 0x44454a58;
    private static final int LABEL   = 0xFF8A8F9C;
    private static final int VALUE   = 0xFF55FF55;
    private static final int GOLD    = 0xFFFFD700;
    private static final int NAME    = 0xFFE8ECF2;
    private static final int BAR_ON  = 0xFF55E05A;
    private static final int BAR_OFF = 0x55202530;

    private static final Pattern COL_KEY  = Pattern.compile("^!([A-Za-z])");
    private static final Pattern SERVER_ID = Pattern.compile("\\b((?:mini|mega|m)\\d+[A-Za-z]{1,3})\\b");

    /**
     * Returns true if the current tab uses Hypixel's lobby column-major encoding
     * (entries named "!A-…"/"!B-…"). Dungeons, Kuudra, Rift, Garden, Crimson Isle
     * sub-servers etc. don't use this — for those we fall back to vanilla rendering
     * via {@link #shouldRender()} so we just draw whatever Hypixel sent verbatim.
     */
    public static boolean shouldRender() {
        Minecraft mc = Minecraft.getInstance();
        if (mc.player == null || mc.getConnection() == null) return false;
        for (PlayerInfo e : mc.getConnection().getOnlinePlayers()) {
            if (COL_KEY.matcher(nameOf(e)).find()) return true;
        }
        return false;
    }

    public static void render(GuiGraphicsExtractor ctx, int screenW, String tabHeader, String tabFooter) {
        Minecraft mc = Minecraft.getInstance();
        if (mc.player == null || mc.getConnection() == null) return;
        Font tr = mc.font;
        int lh = 10;

        // ── group entries into Hypixel's tab columns (by their !X- sort key) ──
        List<PlayerInfo> all = new ArrayList<>(mc.getConnection().getOnlinePlayers());
        all.sort((a, b) -> nameOf(a).compareToIgnoreCase(nameOf(b)));
        Map<String, List<PlayerInfo>> grouped = new LinkedHashMap<>();
        for (PlayerInfo e : all) {
            Matcher m = COL_KEY.matcher(nameOf(e));
            if (m.find()) grouped.computeIfAbsent(m.group(1).toUpperCase(), k -> new ArrayList<>()).add(e);
        }
        // Non-lobby tabs (dungeons / Kuudra / Rift / Garden / etc.) have no !X- keys —
        // caller should have routed to vanilla via shouldRender(); guard anyway.
        if (grouped.isEmpty()) return;

        // ── smart sizing: trim trailing blank rows, drop empty columns, width = content ──
        List<List<PlayerInfo>> columns = new ArrayList<>();
        List<Integer> colWidths = new ArrayList<>();
        int rows = 0;
        boolean first = true;
        for (var col : grouped.values()) {
            int last = -1, nonBlank = 0;
            for (int i = 0; i < col.size(); i++) if (!blank(col.get(i))) { last = i; nonBlank++; }
            // Drop empty columns AND header-only columns (e.g. an "Info" column with no content).
            if (last < 0 || nonBlank <= 1) { first = false; continue; }
            List<PlayerInfo> trimmed = new ArrayList<>(col.subList(0, last + 1));
            boolean playersCol = first; first = false;
            int maxW = 0;
            for (PlayerInfo e : trimmed) { Component dn = e.getTabListDisplayName(); if (dn != null) maxW = Math.max(maxW, tr.width(dn)); }
            int extra = playersCol ? 26 : 8;                       // head + signal bars on players col
            int w = Math.max(playersCol ? 116 : 60, Math.min(maxW + extra, 230));
            columns.add(trimmed);
            colWidths.add(w);
            rows = Math.max(rows, trimmed.size());
        }
        if (columns.isEmpty()) return;
        rows = Math.min(rows, 22);

        int ping = realPing(mc);
        int fps = mc.getFps();
        double tps = fishmod.features.dungeon.PartyCommandHandler.currentTps();
        String server = findServer(mc, tabFooter, tabHeader);

        // ── panel geometry ──
        int pad = 8, gap = 8;
        int contentW = 0; for (int w : colWidths) contentW += w;
        contentW += gap * (columns.size() - 1);
        int pw = Math.min(screenW - 12, contentW + pad * 2);
        int x0 = (screenW - pw) / 2;
        int y0 = 4;
        int headH = 28;
        int bodyH = rows * lh + 6;
        int footH = 12;
        int totalH = headH + bodyH + footH;

        // rounded translucent panel
        roundRect(ctx, x0, y0, x0 + pw, y0 + totalH, bgPanel());
        roundRect(ctx, x0, y0, x0 + pw, y0 + headH, bgHead());

        // ── header stat bar (PLAYERS cell removed — count already shown atop the Players column) ──
        String[] labels = {"SERVER", "TPS", "FPS", "PING"};
        String[] values = {
                server,
                tps < 0 ? "—" : String.format("%.2f", tps),
                String.valueOf(fps),
                ping < 0 ? "—" : ping + "ms"
        };
        int cellW = pw / labels.length;
        for (int i = 0; i < labels.length; i++) {
            int cxL = x0 + i * cellW;
            if (i > 0) ctx.fill(cxL, y0 + 5, cxL + 1, y0 + headH - 5, DIVIDER);
            int cxC = cxL + cellW / 2;
            ctx.centeredText(tr, "§7" + labels[i], cxC, y0 + 5, LABEL);
            int vc = (i == 1 && tps >= 0 && tps < 19) ? 0xFFFF5555 : VALUE;
            ctx.centeredText(tr, values[i], cxC, y0 + 16, vc);
        }
        ctx.fill(x0 + 4, y0 + headH - 1, x0 + pw - 4, y0 + headH, DIVIDER);

        // ── columns ──
        int cy = y0 + headH + 3;
        int colX = x0 + pad;
        for (int c = 0; c < columns.size(); c++) {
            int w = colWidths.get(c);
            if (c > 0) ctx.fill(colX - gap / 2, cy, colX - gap / 2 + 1, cy + rows * lh, DIVIDER);
            boolean playersCol = (c == 0);
            List<PlayerInfo> entries = columns.get(c);
            for (int r = 0; r < entries.size() && r < rows; r++) {
                PlayerInfo e = entries.get(r);
                Component dn = e.getTabListDisplayName();
                if (dn == null) continue;
                int ry = cy + r * lh;
                int tx = colX;
                if (playersCol && r > 0) {
                    try { PlayerFaceExtractor.extractRenderState(ctx, e.getSkin(), colX, ry - 1, 8); } catch (Exception ignored) {}
                    tx = colX + 10;
                }
                // draw styled text directly (no plain-string trim → keeps rank colors)
                ctx.text(tr, dn, tx, ry, NAME, true);
                if (playersCol && r > 0 && e.getLatency() > 0)
                    drawSignal(ctx, colX + w - 13, ry, e.getLatency());
            }
            colX += w + gap;
        }

        ctx.centeredText(tr, footerLine(tabFooter), x0 + pw / 2, y0 + totalH - footH + 2, GOLD);
    }

    private static boolean blank(PlayerInfo e) {
        Component dn = e.getTabListDisplayName();
        if (dn == null) return true;
        // Strip color codes AND invisible formatting chars (NBSP, LRM/RLM, ZWJ, separators) so
        // Hypixel's hidden-character padding rows are recognized as blank.
        String s = dn.getString().replaceAll("§.", "").replaceAll("[\\p{Cf}\\p{Z}\\s]", "");
        return s.isEmpty();
    }

    /** 2px-radius rounded rectangle fill. */
    private static void roundRect(GuiGraphicsExtractor ctx, int x1, int y1, int x2, int y2, int color) {
        ctx.fill(x1 + 2, y1, x2 - 2, y1 + 1, color);
        ctx.fill(x1 + 1, y1 + 1, x2 - 1, y1 + 2, color);
        ctx.fill(x1, y1 + 2, x2, y2 - 2, color);
        ctx.fill(x1 + 1, y2 - 2, x2 - 1, y2 - 1, color);
        ctx.fill(x1 + 2, y2 - 1, x2 - 2, y2, color);
    }

    private static void drawSignal(GuiGraphicsExtractor ctx, int x, int y, int latency) {
        int filled = latency <= 75 ? 4 : latency <= 150 ? 3 : latency <= 300 ? 2 : 1;
        for (int b = 0; b < 4; b++) {
            int h = 2 + b * 2;
            int bx = x + b * 3;
            ctx.fill(bx, y + 8 - h, bx + 2, y + 8, b < filled ? BAR_ON : BAR_OFF);
        }
    }

    private static String nameOf(PlayerInfo e) {
        try { return e.getProfile() != null && e.getProfile().name() != null ? e.getProfile().name() : ""; }
        catch (Exception ex) { return ""; }
    }

    /**
     * Real ping. The vanilla ping/pong round trip ({@link fishmod.utils.PingTracker}) is the most
     * accurate, freshest end-to-end source — a true client→server→client measurement, the same one
     * Odin uses. Fall back to server-measured tab latency, then server-list join ping, only when a
     * live measurement isn't available yet (briefly after join).
     */
    private static int realPing(Minecraft mc) {
        int live = fishmod.utils.PingTracker.latest();
        if (live > 0) return live;
        try {
            var self = mc.getConnection().getPlayerInfo(mc.player.getUUID());
            if (self != null && self.getLatency() > 0) return self.getLatency();
        } catch (Exception ignored) {}
        try {
            var si = mc.getCurrentServer();
            if (si != null && si.ping > 0) return (int) si.ping;
        } catch (Exception ignored) {}
        return -1;
    }

    private static String findServer(Minecraft mc, String footer, String header) {
        String hay = (footer == null ? "" : footer) + " " + (header == null ? "" : header);
        // sidebar often carries the server id (e.g. "05/27/26 m108AB")
        try {
            if (mc.level != null) {
                var sb = mc.level.getScoreboard();
                var obj = sb.getDisplayObjective(net.minecraft.world.scores.DisplaySlot.SIDEBAR);
                if (obj != null) for (var en : sb.listPlayerScores(obj)) {
                    var team = sb.getPlayersTeam(en.owner());
                    String raw = team != null ? team.getPlayerPrefix().getString() + en.owner() + team.getPlayerSuffix().getString() : en.ownerName().getString();
                    hay += " " + raw.replaceAll("§.", "");
                }
            }
        } catch (Exception ignored) {}
        Matcher m = SERVER_ID.matcher(hay);
        return m.find() ? m.group(1) : "—";
    }

    private static String footerLine(String footer) {
        if (footer != null && !footer.isEmpty()) for (String line : footer.split("\n")) {
            String s = line.replaceAll("§.", "").trim();
            if (s.toUpperCase().contains("STORE") || s.toUpperCase().contains("RANKS")) return "§6" + s;
        }
        return "§6Ranks, Boosters & MORE! §eSTORE.HYPIXEL.NET";
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/CompactTab.kt
package fishmod.features

import fishmod.features.dungeon.PartyCommandHandler
import fishmod.utils.PingTracker
import fishmod.utils.TabListCache
import fishmod.utils.config.values.FishSettings
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Font
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.components.PlayerFaceExtractor
import net.minecraft.client.multiplayer.PlayerInfo
import net.minecraft.network.chat.Component
import net.minecraft.world.scores.DisplaySlot
import java.util.regex.Pattern
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

object CompactTab {

    private const val BG_RGB = 0x0B0D13

    private fun bgPanel(): Int {
        val pct = max(0, min(100, FishSettings.compactTabOpacity))
        val a = (pct * 2.55).roundToInt()
        return (a shl 24) or BG_RGB
    }

    private fun bgHead(): Int {
        val pct = max(0, min(100, FishSettings.compactTabOpacity))
        val a = (pct * 2.55 * 0.45).roundToInt()
        return (a shl 24) or BG_RGB
    }

    private const val BORDER = 0x66303440
    private const val DIVIDER = 0x44454a58
    private const val LABEL = 0xFF8A8F9C.toInt()
    private const val VALUE = 0xFF55FF55.toInt()
    private const val GOLD = 0xFFFFD700.toInt()
    private const val NAME = 0xFFE8ECF2.toInt()
    private const val BAR_ON = 0xFF55E05A.toInt()
    private const val BAR_OFF = 0x55202530

    private val COL_KEY: Pattern = Pattern.compile("^!([A-Za-z])")
    private val SERVER_ID: Pattern = Pattern.compile("\\b((?:mini|mega|m)\\d+[A-Za-z]{1,3})\\b")

    // --- Tab Sort ---------------------------------------------------------
    // Purely text-based (no Hypixel API calls): everything is read off the
    // already-rendered tab-list display name string.

    private val SORT_LEVEL_TAG = Regex("""\[(\d{1,4})[^\[\]\d]{0,4}]""")
    private val SORT_RANK_TAG = Regex("""\[([A-Za-z+]{2,10})]""")

    // Highest to lowest. GM and MOD share a tier.
    private val RANK_TIER_GROUPS: List<Set<String>> = listOf(
        setOf("OWNER"),
        setOf("ADMIN"),
        setOf("GM", "MOD"),
        setOf("YOUTUBE"),
        setOf("MVP++"),
        setOf("MVP+"),
        setOf("MVP"),
        setOf("VIP+"),
        setOf("VIP"),
    )

    // Best-effort glyph set for Ironman/Bingo tags Hypixel appends near the
    // player's name. Codepoints unverified without live game data — tune
    // this constant after testing in-game; an empty match just means every
    // player falls through to "unflagged", so it never breaks the sort.
    private val IRONMAN_BINGO_MARKERS = Regex("[☘♻☢⚘🌱]") // ☘ ♻ ☢ ⚚ 🌱

    // Hypixel's usual tab-list relationship colors: gold = party, aqua = friend, dark green = guild.
    private val SOCIAL_COLOR_CODE = Regex("§([0-9a-fk-or])", RegexOption.IGNORE_CASE)
    private val BRACKET_TAG = Regex("""\[[^\[\]]*]""")

    private fun sortLevelOf(stripped: String): Int =
        SORT_LEVEL_TAG.find(stripped)?.groupValues?.get(1)?.toIntOrNull() ?: 0

    private fun sortRankTierOf(stripped: String): Int {
        for (m in SORT_RANK_TAG.findAll(stripped)) {
            val tag = m.groupValues[1].uppercase()
            val idx = RANK_TIER_GROUPS.indexOfFirst { tag in it }
            if (idx >= 0) return idx
        }
        return RANK_TIER_GROUPS.size
    }

    private fun sortIsIronmanBingo(raw: String): Boolean = IRONMAN_BINGO_MARKERS.containsMatchIn(raw)

    private fun sortSocialTierOf(raw: String): Int {
        // Strip bracket contents (rank/level tags) but keep the color codes
        // around them, so the last remaining color code is the one that was
        // applied to the actual username text.
        val withoutTags = BRACKET_TAG.replace(raw, "")
        val code = SOCIAL_COLOR_CODE.findAll(withoutTags).lastOrNull()?.groupValues?.get(1)?.lowercase()
        return when (code) {
            "6" -> 0 // party - gold
            "b" -> 1 // friends - aqua
            "2" -> 2 // guild - dark green
            else -> 3 // everyone else
        }
    }

    // Applies only to the real player-list column (col 0); every other
    // column is a fixed informational panel and must keep its server-given
    // row order regardless of the chosen sort mode. Row 0 of the players
    // column is a fixed header (e.g. "Players (19)"), not a sortable
    // player — it stays pinned in place and drawColumns never gives it a
    // face/ping bar.
    private fun sortPlayersColumn(entries: List<PlayerInfo>): List<PlayerInfo> {
        if (entries.size <= 1) return entries
        val header = entries[0]
        val rest = entries.subList(1, entries.size)
        val mode = FishSettings.compactTabSortMode
        val sortedRest = if (mode == "Random") rest.shuffled() else {
            fun stripped(e: PlayerInfo) = BLANK_COLOR.matcher(e.tabListDisplayName?.string ?: "").replaceAll("")
            fun raw(e: PlayerInfo) = e.tabListDisplayName?.string ?: ""
            val cmp: Comparator<PlayerInfo> = when (mode) {
                "SB Level" -> compareByDescending { sortLevelOf(stripped(it)) }
                "Name (Abc)" -> compareBy<PlayerInfo> { nameOf(it).lowercase() }
                    .thenByDescending { sortLevelOf(stripped(it)) }
                "Ironman/Bingo" -> compareBy<PlayerInfo> { if (sortIsIronmanBingo(raw(it))) 0 else 1 }
                    .thenByDescending { sortLevelOf(stripped(it)) }
                "Party/Friends/Guild" -> compareBy<PlayerInfo> { sortSocialTierOf(raw(it)) }
                    .thenByDescending { sortLevelOf(stripped(it)) }
                else -> compareBy<PlayerInfo> { sortRankTierOf(stripped(it)) } // "Rank (Default)" and unknown values
                    .thenByDescending { sortLevelOf(stripped(it)) }
            }
            rest.sortedWith(cmp)
        }
        val out = ArrayList<PlayerInfo>(entries.size)
        out.add(header)
        out.addAll(sortedRest)
        return out
    }

    private var shouldRenderVersion = -1
    private var shouldRenderCached = false

    @JvmStatic
    fun shouldRender(): Boolean {
        val mc = Minecraft.getInstance()
        if (mc.player == null || mc.connection == null) return false
        val v = TabListCache.version
        if (v == shouldRenderVersion) return shouldRenderCached
        shouldRenderVersion = v
        shouldRenderCached = TabListCache.entries.any { COL_KEY.matcher(nameOf(it.info)).find() }
        return shouldRenderCached
    }

    @JvmStatic
    fun render(ctx: GuiGraphicsExtractor, screenW: Int, tabHeader: String?, tabFooter: String?) {
        val mc = Minecraft.getInstance()
        if (mc.player == null || mc.connection == null) return
        val tr = mc.font
        val lh = 10

        val mdl = model(mc, tabHeader, tabFooter) ?: return
        val columns = mdl.columns
        val colWidths = mdl.colWidths
        val rows = mdl.rows

        val ping = realPing(mc)
        val fps = mc.fps
        val tps = PartyCommandHandler.currentTps()
        val server = mdl.server

        val labels = arrayOf("SERVER", "TPS", "FPS", "PING")
        val values = arrayOf(
            server,
            if (tps < 0) "—" else String.format("%.2f", tps),
            fps.toString(),
            if (ping < 0) "—" else "${ping}ms"
        )
        fun valueColor(i: Int) = if (i == 1 && tps >= 0 && tps < 19) 0xFFFF5555.toInt() else VALUE

        val pad = 8
        val gap = 8
        var contentW = 0
        for (w in colWidths) contentW += w
        contentW += gap * (columns.size - 1)
        val bodyH = rows * lh + 6
        val footH = 12
        val topPad = 8
        val tabW = contentW + pad * 2
        val tabH = topPad + bodyH + footH
        val boxGap = 8

        if (!FishSettings.compactTabStatBarEnabled) {
            val w = min(screenW - 12, tabW)
            val x0 = (screenW - w) / 2
            val y0 = 4
            roundRect(ctx, x0, y0, x0 + w, y0 + tabH, bgPanel())
            drawColumns(ctx, tr, columns, colWidths, x0 + pad, y0 + topPad, rows, gap, lh)
            ctx.centeredText(tr, mdl.footer, x0 + w / 2, y0 + tabH - footH + 2, GOLD)
            return
        }

        val pos = FishSettings.compactTabStatBarPosition.uppercase()
        if (pos == "LEFT" || pos == "RIGHT") {
            val statLineH = 14
            var statW = 0
            for (i in labels.indices) statW = max(statW, tr.width("§7" + labels[i] + " " + values[i]))
            statW += 16
            val statH = tabH

            val totalW = tabW + boxGap + statW
            val w = min(screenW - 12, totalW)
            val x0 = (screenW - w) / 2
            val y0 = 4
            val tabX0 = if (pos == "LEFT") x0 + statW + boxGap else x0
            val statX0 = if (pos == "LEFT") x0 else x0 + tabW + boxGap

            roundRect(ctx, tabX0, y0, tabX0 + tabW, y0 + tabH, bgPanel())
            drawColumns(ctx, tr, columns, colWidths, tabX0 + pad, y0 + topPad, rows, gap, lh)
            ctx.centeredText(tr, mdl.footer, tabX0 + tabW / 2, y0 + tabH - footH + 2, GOLD)

            roundRect(ctx, statX0, y0, statX0 + statW, y0 + statH, bgPanel())
            val cellH = statH / labels.size
            for (i in labels.indices) {
                val cellTop = y0 + i * cellH
                val sy = cellTop + (cellH - statLineH) / 2 + 3
                if (i > 0) ctx.fill(statX0 + 6, cellTop, statX0 + statW - 6, cellTop + 1, DIVIDER)
                ctx.text(tr, "§7" + labels[i] + " ", statX0 + 8, sy, LABEL, false)
                val lw = tr.width("§7" + labels[i] + " ")
                ctx.text(tr, values[i], statX0 + 8 + lw, sy, valueColor(i), false)
            }
        } else {
            val statBarH = 32
            val gapTB = 4
            val w = min(screenW - 12, tabW)
            val x0 = (screenW - w) / 2
            val y0 = 4
            val bottom = pos == "BOTTOM"
            val tabY0 = if (bottom) y0 else y0 + statBarH + gapTB
            val statY0 = if (bottom) tabY0 + tabH + gapTB else y0

            roundRect(ctx, x0, tabY0, x0 + w, tabY0 + tabH, bgPanel())
            drawColumns(ctx, tr, columns, colWidths, x0 + pad, tabY0 + topPad, rows, gap, lh)
            ctx.centeredText(tr, mdl.footer, x0 + w / 2, tabY0 + tabH - footH + 2, GOLD)

            roundRect(ctx, x0, statY0, x0 + w, statY0 + statBarH, bgPanel())
            val cellW = w / labels.size
            for (i in labels.indices) {
                val cxL = x0 + i * cellW
                if (i > 0) ctx.fill(cxL, statY0 + 6, cxL + 1, statY0 + statBarH - 6, DIVIDER)
                val cxC = cxL + cellW / 2
                ctx.centeredText(tr, "§7" + labels[i], cxC, statY0 + 7, LABEL)
                ctx.centeredText(tr, values[i], cxC, statY0 + 18, valueColor(i))
            }
        }
    }

    private class Model(
        val columns: List<List<PlayerInfo>>,
        val colWidths: List<Int>,
        val rows: Int,
        val server: String,
        val footer: String,
    )

    private const val MODEL_TTL_MS = 1000L
    private var modelVersion = -1
    private var modelHeaderFooterSig = ""
    private var modelAt = 0L
    private var cachedModel: Model? = null

    private fun model(mc: Minecraft, tabHeader: String?, tabFooter: String?): Model? {
        val now = System.currentTimeMillis()
        val v = TabListCache.version
        val hf = (tabHeader ?: "") + " " + (tabFooter ?: "")
        val cached = cachedModel
        if (cached != null && v == modelVersion && hf == modelHeaderFooterSig && now - modelAt < MODEL_TTL_MS) return cached
        modelVersion = v
        modelHeaderFooterSig = hf
        modelAt = now
        cachedModel = buildModel(mc, tabHeader, tabFooter)
        return cachedModel
    }

    private fun buildModel(mc: Minecraft, tabHeader: String?, tabFooter: String?): Model? {
        val tr = mc.font
        val all = ArrayList<PlayerInfo>(TabListCache.entries.size)
        for (e in TabListCache.entries) all.add(e.info)
        all.sortWith(Comparator { a, b -> nameOf(a).compareTo(nameOf(b), ignoreCase = true) })
        val grouped = LinkedHashMap<String, MutableList<PlayerInfo>>()
        for (e in all) {
            val m = COL_KEY.matcher(nameOf(e))
            if (m.find()) grouped.getOrPut(m.group(1).uppercase()) { ArrayList() }.add(e)
        }
        if (grouped.isEmpty()) return null

        val columns = ArrayList<List<PlayerInfo>>()
        val colWidths = ArrayList<Int>()
        var rows = 0
        var first = true
        for (col in grouped.values) {
            var last = -1
            var nonBlank = 0
            for (i in col.indices) if (!blank(col[i])) {
                last = i
                nonBlank++
            }
            if (last < 0 || nonBlank <= 1) {
                first = false
                continue
            }
            val playersCol = first
            first = false
            val trimmed = if (playersCol) sortPlayersColumn(col.subList(0, last + 1))
                          else ArrayList(col.subList(0, last + 1))
            var maxW = 0
            for (e in trimmed) {
                val dn = e.tabListDisplayName
                if (dn != null) maxW = max(maxW, tr.width(dn))
            }
            val extra = if (playersCol) 26 else 8
            val w = max(if (playersCol) 116 else 60, min(maxW + extra, 230))
            columns.add(trimmed)
            colWidths.add(w)
            rows = max(rows, trimmed.size)
        }
        if (columns.isEmpty()) return null
        rows = min(rows, 22)

        return Model(columns, colWidths, rows, findServer(mc, tabFooter, tabHeader), footerLine(tabFooter))
    }

    private fun drawColumns(
        ctx: GuiGraphicsExtractor, tr: Font,
        columns: List<List<PlayerInfo>>, colWidths: List<Int>,
        startX: Int, cy: Int, rows: Int, gap: Int, lh: Int
    ) {
        var colX = startX
        for (c in columns.indices) {
            val w = colWidths[c]
            if (c > 0) ctx.fill(colX - gap / 2, cy, colX - gap / 2 + 1, cy + rows * lh, DIVIDER)
            val playersCol = c == 0
            val entries = columns[c]
            var r = 0
            while (r < entries.size && r < rows) {
                val e = entries[r]
                val dn = fishmod.cosmetic.prestige.PrestigeLevelColors.colorizeLevelPrefix(e.tabListDisplayName)
                if (dn == null) {
                    r++
                    continue
                }
                val ry = cy + r * lh
                var tx = colX
                if (playersCol && r > 0) {
                    try {
                        PlayerFaceExtractor.extractRenderState(ctx, e.skin, colX, ry - 1, 8)
                    } catch (ignored: Exception) {
                    }
                    tx = colX + 10
                }
                ctx.text(tr, dn, tx, ry, NAME, true)
                if (playersCol && r > 0 && e.latency > 0) drawSignal(ctx, colX + w - 13, ry, e.latency)
                r++
            }
            colX += w + gap
        }
    }

    private val BLANK_COLOR: Pattern = Pattern.compile("§.")
    private val BLANK_INVISIBLE: Pattern = Pattern.compile("[\\p{Cf}\\p{Z}\\s]")

    private fun blank(e: PlayerInfo): Boolean {
        val dn = e.tabListDisplayName ?: return true
        val s = BLANK_INVISIBLE.matcher(BLANK_COLOR.matcher(dn.string).replaceAll("")).replaceAll("")
        return s.isEmpty()
    }

    private fun roundRect(ctx: GuiGraphicsExtractor, x1: Int, y1: Int, x2: Int, y2: Int, color: Int) {
        ctx.fill(x1 + 2, y1, x2 - 2, y1 + 1, color)
        ctx.fill(x1 + 1, y1 + 1, x2 - 1, y1 + 2, color)
        ctx.fill(x1, y1 + 2, x2, y2 - 2, color)
        ctx.fill(x1 + 1, y2 - 2, x2 - 1, y2 - 1, color)
        ctx.fill(x1 + 2, y2 - 1, x2 - 2, y2, color)
    }

    private fun drawSignal(ctx: GuiGraphicsExtractor, x: Int, y: Int, latency: Int) {
        val filled = if (latency <= 75) 4 else if (latency <= 150) 3 else if (latency <= 300) 2 else 1
        for (b in 0 until 4) {
            val h = 2 + b * 2
            val bx = x + b * 3
            ctx.fill(bx, y + 8 - h, bx + 2, y + 8, if (b < filled) BAR_ON else BAR_OFF)
        }
    }

    private fun nameOf(e: PlayerInfo): String {
        return try {
            if (e.profile != null && e.profile.name != null) e.profile.name else ""
        } catch (ex: Exception) {
            ""
        }
    }

    private fun realPing(mc: Minecraft): Int {
        val live = PingTracker.latest()
        if (live > 0) return live
        try {
            val self = mc.connection?.getPlayerInfo(mc.player!!.uuid)
            if (self != null && self.latency > 0) return self.latency
        } catch (ignored: Exception) {
        }
        try {
            val si = mc.currentServer
            if (si != null && si.ping > 0) return si.ping.toInt()
        } catch (ignored: Exception) {
        }
        return -1
    }

    private fun findServer(mc: Minecraft, footer: String?, header: String?): String {
        var hay = (footer ?: "") + " " + (header ?: "")
        try {
            if (mc.level != null) {
                val sb = mc.level!!.scoreboard
                val obj = sb.getDisplayObjective(DisplaySlot.SIDEBAR)
                if (obj != null) for (en in sb.listPlayerScores(obj)) {
                    val team = sb.getPlayersTeam(en.owner())
                    val raw = if (team != null) team.playerPrefix.string + en.owner() + team.playerSuffix.string else en.ownerName().string
                    hay += " " + raw.replace(Regex("§."), "")
                }
            }
        } catch (ignored: Exception) {
        }
        val m = SERVER_ID.matcher(hay)
        return if (m.find()) m.group(1) else "—"
    }

    private fun footerLine(footer: String?): String {
        if (footer != null && footer.isNotEmpty()) for (line in footer.split("\n")) {
            val s = line.replace(Regex("§."), "").trim()
            if (s.uppercase().contains("STORE") || s.uppercase().contains("RANKS")) return "§6$s"
        }
        return "§6Ranks, Boosters & MORE! §eSTORE.HYPIXEL.NET"
    }
}
